scio-core/src/main/scala/com/spotify/scio/sql/Query.scala

```diff
@@ -207,6 +114,178 @@ object Query {
+      override def query: String = q
+
+      override def apply(a: SCollection[A], b: SCollection[B]): SCollection[Row] =
+        a.context.wrap {
+          val collA = QueryUtils.transform(a)
+          val collB = QueryUtils.transform(b)
+          val sqlTransform = SqlTransform.query(query).registerUdf(udfs: _*)
+
+          PCollectionTuple
+            .of(new TupleTag[A](left), collA.internal)
+            .and(new TupleTag[B](right), collB.internal)
+            .apply(s"${collA.tfName} join ${collB.tfName}", sqlTransform)
+        }
+    }
+
+  def of[A: Schema, B: Schema, O: Schema](q: String,
+                                          left: String,
+                                          right: String,
+                                          udfs: Udf*): Query2[A, B, O] =
+    new Query2[A, B, O] {
+      val query: String = q
+
+      def apply(a: SCollection[A], b: SCollection[B]): SCollection[O] = {
+        try {
+          val (schema, to, from) = SchemaMaterializer.materialize(a.context, Schema[O])
+          row[A, B](query, left, right, udfs: _*)
+            .apply(a, b)
+            .map[O](r => from(r))(Coder.beam(SchemaCoder.of(schema, to, from)))
+        } catch {
+          case e: ParseException =>
+            typecheck(this, left, right)
+              .fold(err => throw new RuntimeException(err, e), _ => throw e)
+        }
+      }
+    }
+
+  def tsql[A: Schema, B: Schema, O: Schema](query: String,
+                                            left: String,
+                                            right: String,
+                                            udfs: Udf*): Query2[A, B, O] =
+    macro QueryMacros.tsqlJoinImpl[A, B, O]
+}
+
+object QueryUtils {
+
+  private[this] val PCollectionName = "PCOLLECTION"
+
+  // Beam is annoyingly verbose when is parses SQL queries.
+  // This function makes is silent.
+  private def silence[A](a: () => A): A = {
+    val prop = "org.slf4j.simpleLogger.defaultLogLevel"
+    val ll = System.getProperty(prop)
+    System.setProperty(prop, "ERROR")
+    val x = a()
+    if (ll != null) System.setProperty(prop, ll)
+    x
+  }
+
+  def transform[T: Schema](c: SCollection[T]): SCollection[T] = {
+    val coderT: Coder[T] = {
+      val (schema, to, from) = SchemaMaterializer.materialize(c.context, Schema[T])
+      Coder.beam(SchemaCoder.of(schema, to, from))
+    }
+    c.transform(s"${c.tfName}: set schema")(_.map(identity)(coderT))
+  }
+
+  def parseQuery(query: String, schemas: (String, BSchema)*): Try[BeamRelNode] = Try {
+    val tables: Map[String, BeamSqlTable] = schemas.map {
+      case (tag, schema) =>
+        tag -> new BaseBeamTable(schema) {
+          override def buildIOReader(begin: PBegin): PCollection[Row] = ???
+
+          override def buildIOWriter(input: PCollection[Row]): POutput = ???
+
+          override def isBounded: PCollection.IsBounded = PCollection.IsBounded.BOUNDED
+        }
+    }.toMap
+
+    silence { () =>
```

Comment:
It's out of the scope of that PR but I think `silence` actually does not work, or a leat not constently. We may remove it.